### PR TITLE
Deposits can be routed only to trusted vaults

### PR DESCRIPTION
**Pull request by pdyraga**
Closes: #95 

Depositors can route their revealed deposits only to trusted vaults and have
trusted vaults notified about new deposits as soon as these deposits get swept.
Vaults not trusted by the Bridge can still be used by Bank balance owners on
their own responsibility - anyone can approve their Bank balance to any
address.

Governance can mark the given vault address as trusted or no longer trusted.
Vaults are not trusted by default.

Trusted vault must meet the following criteria:
- `IVault.onBalanceIncreased` must have a known, low gas cost.
- `IVault.onBalanceIncreased` must never revert.

Without restricting reveal only to trusted vaults, malicious vaults not meeting
the criteria would be able to nuke sweep proof transactions executed by ECDSA
wallet with deposits routed to them.

Additionally, `DepositRevealed` even holds now vault address.

Knowing the vault address, off-chain wallet can group deposits
in sweep batches by vault to minimize the cost of the sweep proof
transaction.

All changes in this PR increased the gas cost of reveal function by approximately
3500 gas units.

### CODE REVIEW

File: solidity/contracts/bridge/Bridge.sol [hunk 1]
```diff
@@ -100,9 +113,26 @@ contract Bridge {
...
-        bytes4 refundLocktime
+        bytes4 refundLocktime,
+        address vault
     );
 
+    /// @notice Allows the Governance to mark the given vault address as trusted
+    ///         or no longer trusted. Vaults are not trusted by default.
+    ///         Trusted vault must meet the following criteria:
+    ///         - `IVault.onBalanceIncreased` must have a known, low gas cost.
+    ///         - `IVault.onBalanceIncreased` must never revert.
```
Comment (tomaszslabon) [inline solidity/contracts/bridge/Bridge.sol:124]: `onBalanceIncreased` will be renamed to `receiveBalanceIncrease` in #110. Let's not forget to change this description?

File: solidity/contracts/bridge/Bridge.sol [hunk 2]
```diff
@@ -100,9 +113,26 @@ contract Bridge {
...
+    /// @dev Without restricting reveal only to trusted vaults, malicious
+    ///      vaults not meeting the criteria would be able to nuke sweep proof
+    ///      transactions executed by ECDSA wallet with  deposits routed to
+    ///      them.
+    /// @param vault The address of the vault
+    /// @param isTrusted flag indicating whether the vault is trusted or not
+    /// @dev Can only be called by the Governance.
+    function setVaultStatus(address vault, bool isTrusted) external onlyOwner {
+        isVaultTrusted[vault] = isTrusted;
```
Comment (tomaszslabon) [inline solidity/contracts/bridge/Bridge.sol:133]: Maybe emit an event?

Comment (pdyraga) [reply solidity/contracts/bridge/Bridge.sol:135]: Good point. Added in https://github.com/keep-network/tbtc-v2/pull/111/commits/f2bff21e9109b3f8c77838548c3092ee1b9e0880.

**Review by lukasz-zimnoch — APPROVED** [review]

File: solidity/contracts/bridge/Bridge.sol (hunk 1 above)
Comment (pdyraga) [reply solidity/contracts/bridge/Bridge.sol:126]: Done in https://github.com/keep-network/tbtc-v2/pull/110/commits/32892c8ebd63dc74b9c54c48da7894c0826f184c.

**Pull request merged**